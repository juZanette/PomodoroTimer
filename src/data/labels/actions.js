import fs from '../../firebase/firebase'

export const setFullscreenDialog = (fullscreenDialog) => ({
  type: 'SET_FULLSCREEN_DIALOG',
  fullscreenDialog,
})

export const setDesktopDialog = (desktopDialog) => ({
  type: 'SET_DESKTOP_DIALOG',
  desktopDialog,
})

export const setDeleteAlert = (deleteAlert) => ({
  type: 'SET_DELETE_ALERT',
  deleteAlert,
})

export const setMenuOpened = (menuOpened) => ({
  type: 'SET_MENU_OPENED',
  menuOpened,
})

export const setFormValue = (formValue) => ({
  type: 'SET_FORM_VALUE',
  formValue,
})

export const setLabelEditting = (labelEditting) => ({
  type: 'SET_LABEL_EDITTING',
  labelEditting,
})

export const setLabelSelected = (labelSelected) => ({
  type: 'SET_LABEL_SELECTED',
  labelSelected,
})

export const addLabel = (label) => ({
  type: 'ADD_LABEL',
  label,
})

export const startAddLabel = (label) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    const ref = await fs.collection(`users/${uid}/labels`).add(label)

    dispatch(
      addLabel({
        id: ref.id,
        ...label,
      })
    )

    return ref
  }
}

export const editLabel = (id, updates) => ({
  type: 'EDIT_LABEL',
  id,
  updates,
})

export const startEditLabel = (id, updates) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    await fs.doc(`users/${uid}/labels/${id}`).update({
      ...updates,
    })

    dispatch(editLabel(id, updates))
  }
}

export const deleteLabel = (id) => ({
  type: 'DELETE_LABEL',
  id,
})

export const startDeleteLabel = (id) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid
    await fs.doc(`users/${uid}/labels/${id}`).delete()

    dispatch(deleteLabel(id))
  }
}

export const setLabels = (labels) => ({
  type: 'SET_LABELS',
  labels,
})

export const startSetLabels = () => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    const labelsRef = await fs.collection(`users/${uid}/labels`).get()

    const labels = labelsRef.docs.map((label) => ({
      id: label.id,
      ...label.data(),
    }))

    dispatch(setLabels(labels))

    return labelsRef
  }
}
