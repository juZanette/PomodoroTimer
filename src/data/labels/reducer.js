const initialState = {
  dialogOpened: false,
  labelToEdit: null,
  formValue: {
    name: null,
    color: null,
  },
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SET_DIALOG_OPENED':
      return {
        ...state,
        dialogOpened: action.dialogOpened,
      }
    case 'SET_LABEL_TO_EDIT':
      return {
        ...state,
        labelToEdit: action.labelToEdit,
      }
    case 'SET_FORM_VALUE':
      return {
        ...state,
        formValue: action.formValue,
      }
    default:
      return state
  }
}
