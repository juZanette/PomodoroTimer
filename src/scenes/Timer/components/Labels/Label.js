import React from 'react'
import styled from 'styled-components'
import Box from '@material-ui/core/Box'
import IconButton from '@material-ui/core/IconButton'
import EditIcon from '@material-ui/icons/Edit'
import DeleteIcon from '@material-ui/icons/Delete'
import grey from '@material-ui/core/colors/grey'
import { useDispatch } from 'react-redux'
import {
  setDialogOpened,
  setLabelToEdit,
  setFormValue,
  setDeleteAlert,
  setFormDialog,
} from '../../data/labels/actions.js'
import { useMediaQuery, useTheme } from '@material-ui/core'

const ColorIndicator = styled.span`
  background: ${({ color }) => color};
  display: inline-block;
  width: 20px;
  height: 20px;
  border-radius: 50%;
  margin-right: 20px;
`

const ActionButton = styled(IconButton)`
  color: ${grey[400]};

  display: ${({ hide }) => (hide ? 'none' : 'inherit')};
`

export const Label = ({ label }) => {
  const theme = useTheme()
  const isMediumScreen = useMediaQuery(theme.breakpoints.up('md'))

  const dispatch = useDispatch()

  const handleEdit = () => {
    dispatch(setLabelToEdit(label))
    dispatch(setFormValue({ name: label.name, color: label.color }))

    if (isMediumScreen) {
      console.log('mali dialog')

      dispatch(setFormDialog(true))
    } else {
      console.log('full dialog')
      dispatch(setDialogOpened(true))
    }
  }

  const handleDelete = () => {
    dispatch(setDeleteAlert({ opened: true, labelToDelete: label }))
  }

  return (
    <Box
      display="flex"
      alignItems="center"
      justifyContent="space-between"
      width="100%"
    >
      <Box display="flex" alignItems="center">
        <ColorIndicator color={label.color}></ColorIndicator>
        <span>{label.name}</span>
      </Box>

      <Box display="flex">
        <ActionButton aria-label="Edit label" onClick={handleEdit}>
          <EditIcon />
        </ActionButton>

        <ActionButton
          hide={!isMediumScreen ? 1 : 0}
          aria-label="Delete label"
          onClick={handleDelete}
        >
          <DeleteIcon />
        </ActionButton>
      </Box>
    </Box>
  )
}
