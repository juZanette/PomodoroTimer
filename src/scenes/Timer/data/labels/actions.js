export const setDialogOpened = (dialogOpened) => ({
  type: 'SET_DIALOG_OPENED',
  dialogOpened,
})

export const setAlertOpened = (alertOpened) => ({
  type: 'SET_ALERT_OPENED',
  alertOpened,
})

export const setLabelToEdit = (labelToEdit) => ({
  type: 'SET_LABEL_TO_EDIT',
  labelToEdit,
})

export const setFormValue = (formValue) => ({
  type: 'SET_FORM_VALUE',
  formValue,
})
