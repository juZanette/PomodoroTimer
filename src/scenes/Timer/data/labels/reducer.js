const initialState = {
  dialogOpened: false,
  formDialogOpened: false,
  deleteAlert: {
    opened: false,
    labelToDelete: null,
  },
  labelToEdit: null,
  formValue: null,
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SET_DIALOG_OPENED':
      return {
        ...state,
        dialogOpened: action.dialogOpened,
      }
    case 'SET_FORM_DIALOG':
      return {
        ...state,
        formDialogOpened: action.formDialogOpened,
      }
    case 'SET_DELETE_ALERT':
      const { opened, labelToDelete } = action.deleteAlert

      return {
        ...state,
        deleteAlert: {
          opened,
          labelToDelete,
        },
      }
    case 'SET_LABEL_TO_EDIT':
      return {
        ...state,
        labelToEdit: action.labelToEdit,
      }
    case 'SET_FORM_VALUE':
      return {
        ...state,
        formValue: action.formValue,
      }
    default:
      return state
  }
}
