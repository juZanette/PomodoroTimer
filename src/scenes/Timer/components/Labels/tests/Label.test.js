import React from 'react'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { createMount } from '@material-ui/core/test-utils'
import { Label, ColorIndicator, ActionButton } from '../Label'
import labels from '../../../../../data/labels/tests/mock-data/labels'

describe('<Label />', () => {
  const mockStore = configureMockStore([thunk])

  let store
  let mount
  let wrapper

  beforeEach(() => {
    store = mockStore()
    store.dispatch = jest.fn()

    mount = createMount()

    wrapper = mount(
      <Provider store={store}>
        <Label label={labels[0]} />
      </Provider>
    )
  })

  afterEach(() => {
    mount.cleanUp()
  })

  test('should render <Label /> correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  test('should display label name', () => {
    const { name } = labels[0]
    expect(wrapper.contains(name)).toBe(true)
  })

  test('should display color indicator with correct color', () => {
    const { color } = labels[0]
    expect(wrapper.find(ColorIndicator).prop('color')).toBe(color)
  })

  test('should call dispatch twice when delete button is clicked', () => {
    wrapper.find(ActionButton).at(1).simulate('click')
    expect(store.dispatch).toHaveBeenCalledTimes(2)
  })

  test('should call dispatch three times when edit button is clicked', () => {
    wrapper.find(ActionButton).at(0).simulate('click')
    expect(store.dispatch).toHaveBeenCalledTimes(3)
  })
})
