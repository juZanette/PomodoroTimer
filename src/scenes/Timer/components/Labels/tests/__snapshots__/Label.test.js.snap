// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Label /> should render <Label /> correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Label
    label={
      Object {
        "color": "#f00",
        "id": "1",
        "name": "Coding",
      }
    }
  >
    <Styled(MuiBox)
      alignItems="center"
      display="flex"
      justifyContent="space-between"
      width="100%"
    >
      <div
        className="MuiBox-root MuiBox-root-2"
      >
        <Styled(MuiBox)
          alignItems="center"
          display="flex"
        >
          <div
            className="MuiBox-root MuiBox-root-3"
          >
            <styled.span
              color="#f00"
            >
              <span
                className="sc-AxjAm dcpLPx"
                color="#f00"
              />
            </styled.span>
            <span>
              Coding
            </span>
          </div>
        </Styled(MuiBox)>
        <Styled(MuiBox)
          display="flex"
        >
          <div
            className="MuiBox-root MuiBox-root-4"
          >
            <Styled(WithStyles(ForwardRef(IconButton)))
              aria-label="Edit label"
              onClick={[Function]}
              size="small"
            >
              <WithStyles(ForwardRef(IconButton))
                aria-label="Edit label"
                className="sc-AxirZ fYxHqd"
                onClick={[Function]}
                size="small"
              >
                <ForwardRef(IconButton)
                  aria-label="Edit label"
                  className="sc-AxirZ fYxHqd"
                  classes={
                    Object {
                      "colorInherit": "MuiIconButton-colorInherit",
                      "colorPrimary": "MuiIconButton-colorPrimary",
                      "colorSecondary": "MuiIconButton-colorSecondary",
                      "disabled": "Mui-disabled",
                      "edgeEnd": "MuiIconButton-edgeEnd",
                      "edgeStart": "MuiIconButton-edgeStart",
                      "label": "MuiIconButton-label",
                      "root": "MuiIconButton-root",
                      "sizeSmall": "MuiIconButton-sizeSmall",
                    }
                  }
                  onClick={[Function]}
                  size="small"
                >
                  <WithStyles(ForwardRef(ButtonBase))
                    aria-label="Edit label"
                    centerRipple={true}
                    className="MuiIconButton-root sc-AxirZ fYxHqd MuiIconButton-sizeSmall"
                    disabled={false}
                    focusRipple={true}
                    onClick={[Function]}
                  >
                    <ForwardRef(ButtonBase)
                      aria-label="Edit label"
                      centerRipple={true}
                      className="MuiIconButton-root sc-AxirZ fYxHqd MuiIconButton-sizeSmall"
                      classes={
                        Object {
                          "disabled": "Mui-disabled",
                          "focusVisible": "Mui-focusVisible",
                          "root": "MuiButtonBase-root",
                        }
                      }
                      disabled={false}
                      focusRipple={true}
                      onClick={[Function]}
                    >
                      <button
                        aria-label="Edit label"
                        className="MuiButtonBase-root MuiIconButton-root sc-AxirZ fYxHqd MuiIconButton-sizeSmall"
                        disabled={false}
                        onBlur={[Function]}
                        onClick={[Function]}
                        onDragLeave={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                        onMouseDown={[Function]}
                        onMouseLeave={[Function]}
                        onMouseUp={[Function]}
                        onTouchEnd={[Function]}
                        onTouchMove={[Function]}
                        onTouchStart={[Function]}
                        tabIndex={0}
                        type="button"
                      >
                        <span
                          className="MuiIconButton-label"
                        >
                          <ForwardRef>
                            <WithStyles(ForwardRef(SvgIcon))>
                              <ForwardRef(SvgIcon)
                                classes={
                                  Object {
                                    "colorAction": "MuiSvgIcon-colorAction",
                                    "colorDisabled": "MuiSvgIcon-colorDisabled",
                                    "colorError": "MuiSvgIcon-colorError",
                                    "colorPrimary": "MuiSvgIcon-colorPrimary",
                                    "colorSecondary": "MuiSvgIcon-colorSecondary",
                                    "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                                    "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                                    "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                                    "root": "MuiSvgIcon-root",
                                  }
                                }
                              >
                                <svg
                                  aria-hidden={true}
                                  className="MuiSvgIcon-root"
                                  focusable="false"
                                  viewBox="0 0 24 24"
                                >
                                  <path
                                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                                  />
                                </svg>
                              </ForwardRef(SvgIcon)>
                            </WithStyles(ForwardRef(SvgIcon))>
                          </ForwardRef>
                        </span>
                        <WithStyles(memo)
                          center={true}
                        >
                          <ForwardRef(TouchRipple)
                            center={true}
                            classes={
                              Object {
                                "child": "MuiTouchRipple-child",
                                "childLeaving": "MuiTouchRipple-childLeaving",
                                "childPulsate": "MuiTouchRipple-childPulsate",
                                "ripple": "MuiTouchRipple-ripple",
                                "ripplePulsate": "MuiTouchRipple-ripplePulsate",
                                "rippleVisible": "MuiTouchRipple-rippleVisible",
                                "root": "MuiTouchRipple-root",
                              }
                            }
                          >
                            <span
                              className="MuiTouchRipple-root"
                            >
                              <TransitionGroup
                                childFactory={[Function]}
                                component={null}
                                exit={true}
                              />
                            </span>
                          </ForwardRef(TouchRipple)>
                        </WithStyles(memo)>
                      </button>
                    </ForwardRef(ButtonBase)>
                  </WithStyles(ForwardRef(ButtonBase))>
                </ForwardRef(IconButton)>
              </WithStyles(ForwardRef(IconButton))>
            </Styled(WithStyles(ForwardRef(IconButton)))>
            <Styled(WithStyles(ForwardRef(IconButton)))
              aria-label="Delete label"
              hide={1}
              onClick={[Function]}
              size="small"
            >
              <WithStyles(ForwardRef(IconButton))
                aria-label="Delete label"
                className="sc-AxirZ faiIQu"
                hide={1}
                onClick={[Function]}
                size="small"
              >
                <ForwardRef(IconButton)
                  aria-label="Delete label"
                  className="sc-AxirZ faiIQu"
                  classes={
                    Object {
                      "colorInherit": "MuiIconButton-colorInherit",
                      "colorPrimary": "MuiIconButton-colorPrimary",
                      "colorSecondary": "MuiIconButton-colorSecondary",
                      "disabled": "Mui-disabled",
                      "edgeEnd": "MuiIconButton-edgeEnd",
                      "edgeStart": "MuiIconButton-edgeStart",
                      "label": "MuiIconButton-label",
                      "root": "MuiIconButton-root",
                      "sizeSmall": "MuiIconButton-sizeSmall",
                    }
                  }
                  hide={1}
                  onClick={[Function]}
                  size="small"
                >
                  <WithStyles(ForwardRef(ButtonBase))
                    aria-label="Delete label"
                    centerRipple={true}
                    className="MuiIconButton-root sc-AxirZ faiIQu MuiIconButton-sizeSmall"
                    disabled={false}
                    focusRipple={true}
                    hide={1}
                    onClick={[Function]}
                  >
                    <ForwardRef(ButtonBase)
                      aria-label="Delete label"
                      centerRipple={true}
                      className="MuiIconButton-root sc-AxirZ faiIQu MuiIconButton-sizeSmall"
                      classes={
                        Object {
                          "disabled": "Mui-disabled",
                          "focusVisible": "Mui-focusVisible",
                          "root": "MuiButtonBase-root",
                        }
                      }
                      disabled={false}
                      focusRipple={true}
                      hide={1}
                      onClick={[Function]}
                    >
                      <button
                        aria-label="Delete label"
                        className="MuiButtonBase-root MuiIconButton-root sc-AxirZ faiIQu MuiIconButton-sizeSmall"
                        disabled={false}
                        hide={1}
                        onBlur={[Function]}
                        onClick={[Function]}
                        onDragLeave={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                        onMouseDown={[Function]}
                        onMouseLeave={[Function]}
                        onMouseUp={[Function]}
                        onTouchEnd={[Function]}
                        onTouchMove={[Function]}
                        onTouchStart={[Function]}
                        tabIndex={0}
                        type="button"
                      >
                        <span
                          className="MuiIconButton-label"
                        >
                          <ForwardRef>
                            <WithStyles(ForwardRef(SvgIcon))>
                              <ForwardRef(SvgIcon)
                                classes={
                                  Object {
                                    "colorAction": "MuiSvgIcon-colorAction",
                                    "colorDisabled": "MuiSvgIcon-colorDisabled",
                                    "colorError": "MuiSvgIcon-colorError",
                                    "colorPrimary": "MuiSvgIcon-colorPrimary",
                                    "colorSecondary": "MuiSvgIcon-colorSecondary",
                                    "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                                    "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                                    "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                                    "root": "MuiSvgIcon-root",
                                  }
                                }
                              >
                                <svg
                                  aria-hidden={true}
                                  className="MuiSvgIcon-root"
                                  focusable="false"
                                  viewBox="0 0 24 24"
                                >
                                  <path
                                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                                  />
                                </svg>
                              </ForwardRef(SvgIcon)>
                            </WithStyles(ForwardRef(SvgIcon))>
                          </ForwardRef>
                        </span>
                        <WithStyles(memo)
                          center={true}
                        >
                          <ForwardRef(TouchRipple)
                            center={true}
                            classes={
                              Object {
                                "child": "MuiTouchRipple-child",
                                "childLeaving": "MuiTouchRipple-childLeaving",
                                "childPulsate": "MuiTouchRipple-childPulsate",
                                "ripple": "MuiTouchRipple-ripple",
                                "ripplePulsate": "MuiTouchRipple-ripplePulsate",
                                "rippleVisible": "MuiTouchRipple-rippleVisible",
                                "root": "MuiTouchRipple-root",
                              }
                            }
                          >
                            <span
                              className="MuiTouchRipple-root"
                            >
                              <TransitionGroup
                                childFactory={[Function]}
                                component={null}
                                exit={true}
                              />
                            </span>
                          </ForwardRef(TouchRipple)>
                        </WithStyles(memo)>
                      </button>
                    </ForwardRef(ButtonBase)>
                  </WithStyles(ForwardRef(ButtonBase))>
                </ForwardRef(IconButton)>
              </WithStyles(ForwardRef(IconButton))>
            </Styled(WithStyles(ForwardRef(IconButton)))>
          </div>
        </Styled(MuiBox)>
      </div>
    </Styled(MuiBox)>
  </Label>
</Provider>
`;
